### PR TITLE
Element.NET - SF-473 - Incorrect port constraint identifier cause IdentifierNotFound for declaration

### PR DESCRIPTION
**Pull request by Craig-J**
- Add tests to cover lookup recursing when finding an error when it
shouldnt
- Fix bug where lookup recurses when finding an error instead of
returning the error
- Fix bug where identifiers starting with a keyword would
cause parse error e.g. constraintBlah
- Fix bug where blocks resolved with captures did not include the
captures in their member list (i.e. function calls)

### CODE REVIEW

File: Element.NET/!1-AST/Declarations/ConstraintDeclaration.cs [hunk 1]
```diff
@@ -5,7 +5,7 @@ namespace Element.AST
     public class IntrinsicConstraintDeclaration : Declaration
     {
         protected override string IntrinsicQualifier { get; } = "intrinsic";
-        protected override string Qualifier { get; } = "constraint";
+        protected override string Qualifier { get; } = "constraint ";
```
Comment (Zephilinox) [inline Element.NET/!1-AST/Declarations/ConstraintDeclaration.cs:8]: I assume you also need to change "function" in FunctionDeclaration.cs?

File: Laboratory/Tests/L0-Parsing/Declarations/Identifiers/IdentifiersStartingWithKeywords-pass.ele [hunk 2]
```diff
@@ -0,0 +1,2 @@
+constraintBlah = 5
+structFoo = 10
```
Comment (Zephilinox) [inline Laboratory/Tests/L0-Parsing/Declarations/Identifiers/IdentifiersStartingWithKeywords-pass.ele:2]: intrinsic functionZoop?

File: Laboratory/Tests/L2-Semantics/_2_TypedFunctions.ele [hunk 3]
```diff
@@ -17,4 +17,6 @@ intrinsic function add(a:Num, b:Num):Num
 addFive(a:Num):Num = add(a, 5)
 
 literalNumNotAConstraint(a:5):10 = a
-literalFunctionNotAConstraint(a:onlyNum):returnsNum = a
\ No newline at end of file
+literalFunctionNotAConstraint(a:onlyNum):returnsNum = a
+
+constraintLookupError(a:ConstraintThatDoesntExist) = a
```
Comment (Zephilinox) [inline Laboratory/Tests/L2-Semantics/_2_TypedFunctions.ele:22]: should this have a return constraint so it's a boundary function for runtime tests?

**Review by Zephilinox — CHANGES_REQUESTED** [review]

File: Laboratory/Tests/L0-Parsing/Declarations/Identifiers/IdentifiersStartingWithKeywords-pass.ele (hunk 2 above)
Comment (Craig-J) [reply Laboratory/Tests/L0-Parsing/Declarations/Identifiers/IdentifiersStartingWithKeywords-pass.ele:2]: Added more tests for `intrinsic` and all the valid combinations

File: Laboratory/Tests/L2-Semantics/_2_TypedFunctions.ele (hunk 3 above)
Comment (Craig-J) [reply Laboratory/Tests/L2-Semantics/_2_TypedFunctions.ele:22]: While we need to add runtime tests for semantics stage - this particular one can't be runtime tested because the boundary constraints can't be serializable types. There's also no difference between compile-time and runtime in a case where the error should be reported at compile time.

Comment (Zephilinox) [reply Laboratory/Tests/L2-Semantics/_2_TypedFunctions.ele:22]: I think this is a difference between the two compilers then, in libelement I'm fairly sure it will first check if it's a boundary function and give an error about that before it tries to do constraint checking.

Comment (Craig-J) [reply Laboratory/Tests/L2-Semantics/_2_TypedFunctions.ele:22]: This doesn't need to be checked as a boundary function because it's a compile-time error to have an unresolved constraint. It's not possible to determine if a function is boundary-compatible until the constraints are all evaluated.

Comment (Zephilinox) [reply Laboratory/Tests/L2-Semantics/_2_TypedFunctions.ele:22]: it's possible just by checking that there is a type annotation, but that's fair, we can leave it for now :)

**Review by Zephilinox — APPROVED** [review]
